### PR TITLE
Implement contract memory access and debug formatting

### PR DESCRIPTION
**Pull request by graydon**
This addresses two sets of related problems we've been discussing recently:

  - How to get contract-memory slices into binaries when `Env=Host` (call this "native" or "test contract", however you like): #179
  - How to get richer debug logging from contracts _at least_ (perhaps only) when building `Env=Host` (multiple args, format strings, static string args): #229, #181, #205 

The common problem / issue here (and what took so long, I tried a lot of different unsatisfying approaches) is "what's the safest and least-terrible way of receiving pointers-into-contract-memory on the host when `Env=Host`". We know what to do when `Env=Guest` -- a contract pointer is a linear memory address, a u32 -- but when `Env=Host` it means we're sharing a process with the host and we might find it ugly, unsafe or even unsound to try to transmit pointers through the narrow vocabulary of `Env`, which only knows about `u64` and `RawVal`.

The answer I came up with eventually is "provide the pointer-passing interfaces in `EnvBase`, have them take lifetime-qualified Rust slices, and _decay those to `Env` calls_ only in the `Env=Guest` case, and even then optionally (eg. the logging interface decays them to no-ops because we want `Env=Guest` builds to be mimimum-sized)". Then make the pointer-receiving paths on `Env` fail when called from a non-VM frame (i.e. native/test -- "definitely not `Guest`"), as they do today.

I think this sorts out how to move forwards with such questions, and more concretely it gives a pattern (as well as a few initial functions) for debug logging functions that take multiple args from _either_ guest or host side, formatting them lazily on print-out. I _think_ these should handle the error-reporting cases people are stuck on (eg. the rich type-conversion errors and such).

### CODE REVIEW

**Comment by graydon** [comment]
I'll add a few more debug-message-logging functions later on, just wanted to post this for an initial glance now.

File: stellar-contract-env-common/src/env.rs [hunk 1]
```diff
@@ -63,6 +106,8 @@ macro_rules! call_macro_with_all_host_functions {
             // argument list of comma-separated arg:type pairs
 
             mod context "x" {
+                // This one variant of logging does not take a format string and
+                // is live in both Env=Guest and Env=Host configurations.
                 {"_", fn log_value(v:RawVal) -> RawVal }
```
Comment (leighmcculloch) [inline stellar-contract-env-common/src/env.rs:111]: Should we get rid of this one?

File: stellar-contract-env-guest/src/guest.rs [hunk 2]
```diff
@@ -4,11 +4,60 @@
...
 
 // In guest code the environment is global/implicit, so is represented as a unit struct.
 #[derive(Copy, Clone, Default)]
 pub struct Guest;
 
+// The Guest struct is only meaningful when compiling for the WASM target. All
+// these fns should not be called at all because the SDK's choice of Env should be
+// Host for a non-WASM build.
+#[cfg(not(target_family = "wasm"))]
+impl EnvBase for Guest {
```
Comment (leighmcculloch) [inline stellar-contract-env-guest/src/guest.rs:18]: ❗ We only build the guest crate as a whole for wasm. When will this crate be used in non-wasm?

File: stellar-contract-env-host/src/host.rs [hunk 3]
```diff
@@ -885,6 +921,81 @@ impl EnvBase for Host {
...
+        // This is only called from native contracts, either when testing or
+        // when the contract is otherwise linked into the same address space as
+        // us. We therefore access the memory we were passed directly.
+        //
+        // This is also why we _panic_ on errors in here, rather than attempting
+        // to return a recoverable error code: native contracts that call this
+        // function do so through APIs that _should_ never pass bad data.
+        //
+        // TODO: we may revisit this choice of panicing in the future, depending
+        // on how we choose to try to contain panics-caused-by-native-contracts.
```
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:935]: +1 I think we should revisit this.

**Review by leighmcculloch — COMMENTED** [review]
Looks good to me. I agree with the TODO you added that we should revisit the panicking. We should at least discuss it in greater depth when more people are available to do so. I have one concern marked with the ❗, and I think my questions in that comment probably indicate I have a poor understanding of how some of this works.

File: stellar-contract-env-guest/src/guest.rs (hunk 2 above)
Comment (graydon) [reply stellar-contract-env-guest/src/guest.rs:18]: Should never be. This is a just-in-case. If you like I can remove it.

Comment (graydon) [reply stellar-contract-env-guest/src/guest.rs:18]: (Note there are static asserts that we're on a 32-bit platform in the other case, so like the crate won't build for most hosts anyways, I don't know if maybe we want it to build-but-be-inert in some other case?)

Comment (leighmcculloch) [reply stellar-contract-env-guest/src/guest.rs:18]: I think it's fine to remove this. The assumption is the entire crate is wasm32 only. Even in the SDK we never import it for non-wasm builds.

Comment (leighmcculloch) [reply stellar-contract-env-guest/src/guest.rs:18]: Oh, although maybe the crate builds for non-wasm in local builds of this workspace. I don't mind if it's included.

**Review by leighmcculloch — APPROVED** [review]

**Review by jayz22 — APPROVED** [review]
This is great, thanks for doing! I'll need more time to digest the nuances but approving first.

File: stellar-contract-env-host/src/events.rs [hunk 4]
```diff
@@ -32,56 +34,85 @@ impl Events {
...
+            None => {
+                for arg in self.args.iter() {
+                    write!(f, "{}", arg)?;
                 }
-                first = false;
-                write!(f, "{:?}", rv)?;
+                Ok(())
+            }
+            Some(fmt) => {
+                let args = dyn_fmt::Arguments::new(fmt, self.args.as_slice());
```
Comment (jayz22) [inline stellar-contract-env-host/src/events.rs:95]: This is really cool, didn't know it!

**Pull request merged**